### PR TITLE
SAAS-3129: Added Instance Chat Transcripts Encrypted Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

**Review by dipsubha06 — APPROVED** [review]

File: collectors/aws/collector.js [hunk 1]
```diff
@@ -1045,7 +1045,12 @@ var postcalls = [
                 reliesOnService: 'connect',
                 reliesOnCall: 'listInstances',
                 override: true
-            }
+            },
+            listInstanceChatTranscriptsStorageConfigs: {
```
Comment (AkhtarAmir) [inline collectors/aws/collector.js:1049]: Can you change this to listInstanceChatTranscriptStorageConfigs?

File: exports.js [hunk 2]
```diff
@@ -411,7 +411,9 @@ module.exports = {
       
         'networkMemberDataEncrypted'    : require(__dirname + '/plugins/aws/managedblockchain/networkMemberDataEncrypted.js'),
 
-        'instanceCallRecordingEncrypted': require(__dirname + '/plugins/aws/connect/instanceCallRecordingEncrypted.js')
+        'instanceCallRecordingEncrypted': require(__dirname + '/plugins/aws/connect/instanceCallRecordingEncrypted.js'),
+        
+        'instanceChatTranscriptsEncrypted': require(__dirname + '/plugins/aws/connect/instanceChatTranscriptsEncrypted.js')
```
Comment (AkhtarAmir) [inline exports.js:416]: We need to shorten this name

File: plugins/aws/connect/instanceChatTranscriptsEncrypted.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Connect Instance Chat Transcripts Encrypted',
+    category: 'Connect',
+    domain: 'Content Delivery',
+    description: 'Ensure that Amazon Connect instances have encryption enabled for Chat Transcripts being saved on S3.',
```
Comment (AkhtarAmir) [inline plugins/aws/connect/instanceChatTranscriptsEncrypted.js:8]:
```suggestion
    description: 'Ensure that Amazon Connect instances have encryption enabled for chat transcripts being saved on S3.',
```

File: plugins/aws/connect/instanceChatTranscriptsEncrypted.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                    !listInstanceChatTranscriptsStorageConfigs.data.StorageConfigs) {
+                    helpers.addResult(results, 3,
+                        `Unable to describe Connect instance chat transcripts storage config: ${helpers.addError(listInstanceChatTranscriptsStorageConfigs)}`,
+                        region, resource);
+                    continue;
+                }
+
+                if (!listInstanceChatTranscriptsStorageConfigs.data.StorageConfigs.length) {
+                    helpers.addResult(results, 0,
+                        'Connect instance does not have any storage config for chat transcriptss',
```
Comment (AkhtarAmir) [inline plugins/aws/connect/instanceChatTranscriptsEncrypted.js:79]:
```suggestion
                        'Connect instance does not have any storage config for chat transcripts',
```

File: plugins/aws/connect/instanceChatTranscriptsEncrypted.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+                        var currentEncryptionLevelString = helpers.ENCRYPTION_LEVELS[currentEncryptionLevel];
+
+                        if (currentEncryptionLevel >= desiredEncryptionLevel) {
+                            helpers.addResult(results, 0,
+                                `Connect instance is using ${currentEncryptionLevelString} for chat transcripts encryption\
+                                which is greater than or equal to the desired encryption level ${config.desiredEncryptionLevelString}`,
+                                region, resource);
+                        } else {
+                            helpers.addResult(results, 2,
+                                `Connect instance is using ${currentEncryptionLevelString} for chat transcripts encryption\
```
Comment (AkhtarAmir) [inline plugins/aws/connect/instanceChatTranscriptsEncrypted.js:113]:
```suggestion
                                `Connect instance is using ${currentEncryptionLevelString} for chat transcripts encryption \
```

File: plugins/aws/connect/instanceChatTranscriptsEncrypted.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                            continue;
+                        }
+    
+                        currentEncryptionLevel = helpers.getEncryptionLevel(describeKey.data.KeyMetadata, helpers.ENCRYPTION_LEVELS);
+
+                        var currentEncryptionLevelString = helpers.ENCRYPTION_LEVELS[currentEncryptionLevel];
+
+                        if (currentEncryptionLevel >= desiredEncryptionLevel) {
+                            helpers.addResult(results, 0,
+                                `Connect instance is using ${currentEncryptionLevelString} for chat transcripts encryption\
```
Comment (AkhtarAmir) [inline plugins/aws/connect/instanceChatTranscriptsEncrypted.js:108]:
```suggestion
                                `Connect instance is using ${currentEncryptionLevelString} for chat transcripts encryption \
```

File: plugins/aws/connect/instanceChatTranscriptsEncrypted.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+                                region, resource);
+                        } else {
+                            helpers.addResult(results, 2,
+                                `Connect instance is using ${currentEncryptionLevelString} for chat transcripts encryption\
+                                which is less than the desired encryption level ${config.desiredEncryptionLevelString}`,
+                                region, resource);
+                        }
+                    } else {
+                        helpers.addResult(results, 2,
+                            'Connect instance does not have encryption enabled for chat transcriptss',
```
Comment (AkhtarAmir) [inline plugins/aws/connect/instanceChatTranscriptsEncrypted.js:119]:
```suggestion
                            'Connect instance does not have encryption enabled for chat transcripts',
```

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/connect/instanceTranscriptsEncrypted.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    domain: 'Content Delivery',
+    description: 'Ensure that Amazon Connect instances have encryption enabled for chat transcripts being saved on S3.',
+    more_info: 'You can configure Amazon Connect instance to save transcripts for chats to be saved on S3. When you save ' +
+        'such data on S3, enable encryption for the data and use a KMS key with desired encrypted level to meet regulatory compliance requirements within your organization.',
+    link: 'https://docs.aws.amazon.com/connect/latest/adminguide/encryption-at-rest.html',
+    recommended_action: 'Modify Connect instance data storage configuration and enable encryption for chat transcripts',
+    apis: ['Connect:listInstances', 'Connect:listInstanceChatTranscriptStorageConfigs', 'KMS:listKeys', 'KMS:describeKey'],
+    settings: {
+        connect_chat_transcripts_encryption_level: {
+            name: 'Connect Chat Transcripts Encryption Level',
```
Comment (AkhtarAmir) [inline plugins/aws/connect/instanceTranscriptsEncrypted.js:16]:
```suggestion
            name: 'Connect Chat Transcripts Target Encryption Level',
```

**Review by AkhtarAmir — APPROVED** [review]

**Pull request merged**